frontend/src/components/smart-contracts/TransactionModal.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { useEffect, useState, Dispatch, SetStateAction } from 'react';
+import { ContractTransaction } from 'ethers';
+import { Modal } from 'components/ui/Modal';
+import { ethers } from 'ethers';
+
+export type CallContractCallback = () => Promise<ContractTransaction>;
+
+export type OnConfirmedCallback = (
+  receipt: ethers.providers.TransactionReceipt
+) => void;
+
+export type TransactionError = {
+  message: string;
+};
+
+export enum TXStatus {
+  Ready = 'Ready',
+  Waiting = 'Waiting',
+  Finished = 'Finished',
+  Error = 'Error',
+}
+
+export const TransactionModal = ({
+  title,
+  show,
+  callContract,
+  onConfirmed,
+  updateState,
+  onFinish,
+  onError,
+}: {
+  title: string;
+  show: boolean;
+  callContract: CallContractCallback;
+  onConfirmed?: OnConfirmedCallback;
+  updateState?: Dispatch<SetStateAction<TXStatus>>;
+  onFinish: () => void;
+  onError?: (arg0: string) => void;
+}) => {
+  const [txStatus, setTxStatus] = useState<TXStatus>(TXStatus.Ready);
+
+  const [errorMessage, setErrorMessage] = useState<string | undefined>(
+    undefined
+  );
+
+  // this is a state machine that cycles through Ready, Waiting, Error (optionally) and Finished states
+  useEffect(() => {
```

Comment:
Can all this go in a hook? It would be nice to have something like this at the component level:

`const { ready, waiting, error } = useTransaction()`

It's hard to reason about what's going on here.

Reply:
It could go into a hook, but I think this might be the only place that hook would ever be used.  The entire purpose of TransactionModal is to encapsulate running a transaction.

I'm not sure which way is more complex.

Reply:
I refactored this component to make it clear what it is doing.  I tried moving the finite state machine functionality into a react hook, but it ended up being just a state, setState with a callback.  I can't think of a way to refactor this any further.